File: Heap (Priority Queue)/46. Permutations/solution.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res, row = [], []

        def backtrack():
            if len(row) == n:
                res.append(row.copy())
                return
            
            for x in nums:
                if x not in row:
                    row.append(x)
                    backtrack()
                    row.pop()
        
        backtrack()
        return res 
    

if __name__ == '__main__':
    """ Example 1: """
    nums = [1,2,3]
    print(Solution().permute(nums))

File: Heap (Priority Queue)/46. Permutations/solution.swift

class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
        
        
        let n = nums.count
        var ans : [[Int]] = []
        var sol : [Int] = []
        
        func backtrack() {
            
            if sol.count == n {
                let newSol = sol
                ans.append(newSol)
                return
            }
            
            for x in nums {
                if !sol.contains(x) {
                    sol.append(x)
                    backtrack()
                    sol.popLast()
                }
            }
            
        }
        
        backtrack()
        
        return ans
    }
}

let s = Solution()

let nums : [Int] = [1, 2,3]
print(s.permute(nums))
